fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -0,0 +1,1143 @@
+		while (!ctx->kmsUrlHeap.empty()) {
+			state std::shared_ptr<KmsUrlCtx> curUrl = ctx->kmsUrlHeap.top();
+			ctx->kmsUrlHeap.pop();
+			tempStack.push(curUrl);
+
+			try {
+				std::string kmsEncryptionFullUrl = getEncryptionFullUrl(curUrl->url);
+				TraceEvent("FetchEncryptionKeyByDomainId_Start", ctx->uid)
+				    .detail("KmsEncryptionFullUrl", Traceable<std::string>::toString(kmsEncryptionFullUrl));
+				Reference<HTTP::Response> _resp = wait(ctx->restClient.doPost(kmsEncryptionFullUrl, requestBody));
+				resp = _resp;
+				curUrl->nRequests++;
+
+				try {
+					parseKmsResponse(ctx, resp, reply.arena, reply.cipherKeyDetails);
+
+					// Push urlCtx back on the ctx->urlHeap
+					while (!tempStack.empty()) {
+						ctx->kmsUrlHeap.emplace(tempStack.top());
+						tempStack.pop();
+					}
+
+					TraceEvent("FetchEncryptionKeyByDomainId_Success", ctx->uid)
+					    .detail("KmsUrl", Traceable<std::string>::toString(curUrl->url));
+					return reply;
+				} catch (Error& e) {
+					TraceEvent("FetchEncryptionKeyByDomainId_RespParseFailure").error(e);
+					curUrl->nResponseParseFailures++;
+					// attempt to fetch encryption details from next KmsUrl
+				}
+			} catch (Error& e) {
+				TraceEvent("FetchEncryptionKeyByDomainId_Failed", ctx->uid).error(e);
+				curUrl->nFailedResponses++;
+				// attempt to fetch encryption details from next KmsUrl
+			}
+		}
+
+		if (pass == 1) {
+			// Re-discover KMS Urls and re-attempt to fetch the encryption key details
+			wait(discoverKmsUrls(ctx));
+		}
+	}
+
+	// Failed to fetch encryption keys from remote KmsUrls.
+	throw encrypt_keys_fetch_failed();
+}
+
+void procureValidationTokensFromFiles(Reference<RESTKmsConnectorCtx> ctx, StringRef details) {
+	if (details.empty()) {
+		TraceEvent("ValidationToken_EmptyFileDetails", ctx->uid).log();
+		throw operation_failed();
+	}
+
+	TraceEvent("ValidationToken", ctx->uid).detail("DetailsStr", Traceable<std::string>::toString(details.toString()));
+
+	std::unordered_map<std::string, std::string> tokenFilePathMap;
+	while (!details.empty()) {
+		StringRef name = details.eat(":");
+		if (name.empty()) {
+			break;
+		}
+		StringRef path = details.eat(",");
+		if (path.empty()) {
+			TraceEvent("ValidationToken_FileDetailsMalformed", ctx->uid)
+			    .detail("FileDetails", Traceable<std::string>::toString(details.toString()));
+			throw operation_failed();
+		}
+
+		tokenFilePathMap.emplace(name.toString(), path.toString());
+		TraceEvent("ValidationToken", ctx->uid)
+		    .detail("FName", Traceable<std::string>::toString(name.toString()))
+		    .detail("Path", Traceable<std::string>::toString(path.toString()));
+	}
+
+	// Clear existing cached validation tokens
+	ctx->validationTokens.clear();
+
+	// Enumerate all token files and extract details
+	uint64_t tokensPayloadSize = 0;
+	for (const auto& item : tokenFilePathMap) {
+		const std::string& tokenName = item.first;
+		const std::string& tokenFile = item.second;
+		std::ifstream ifs;
```

Comment:
Same comment as above, why not use flow file apis?

Reply:
I wasn't aware of them, thanks for pointing out. Will take a look.